ISPN-13495 IRAC send tombstones in state transfer

pruivo opened the pull request:
Two changes are made in the commit:

* all tombstones are sent to the new owner.
* backup owner ping primary owner if there is tombstone stored without any update in IracManager.

https://issues.redhat.com/browse/ISPN-13495

offtopic, updated python script to python3

danberindei commented on core/src/test/java/org/infinispan/xsite/irac/IracTombstoneCleanupTest.java:
```diff
@@ -0,0 +1,376 @@
+      return (DefaultIracTombstoneManager) tombstoneManager;
+   }
+
+   private RecordingRpcManager rpcManager(Cache<?, ?> cache) {
```
```suggestion
   private RecordingRpcManager recordingRpcManager(Cache<?, ?> cache) {
```

danberindei commented on core/src/test/java/org/infinispan/xsite/irac/IracTombstoneCleanupTest.java:
```diff
@@ -0,0 +1,376 @@
+      assertEquals(key, cmd.getTombstone().getKey());
+      assertEquals(metadata, cmd.getTombstone().getMetadata());
+
+      pRpcManager.stopRecording();
```
I suggest doing this in an `@AfterMethod(alwaysRun=true)` method instead, so you don't have to worry about a failure in one test making the next test fail as well.

danberindei reviewed (commented):
I don't have a lot of concrete suggestions, but I'm slightly concerned about the lack of backpressure when sending tombstones.

pruivo commented on core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java:
```diff
@@ -117,73 +151,104 @@ public int size() {
+         cmd.addTombstone(tombstone);
+         if (++currentBatch == batchSize) {
+            // send and create a new command
+            rpcManager.sendTo(origin, cmd, DeliverOrder.NONE);
```
agree. I didn't think about backpressure at all (I'm not sure if it is going to grow that much TBH).
so, N=1?

danberindei replied:
Yep, N=1 is good, and we can scale up later if needed

pruivo commented on core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java:
```diff
@@ -117,73 +151,104 @@ public int size() {
    }
 
+   @Override
+   public void sendStateTo(Address origin, IntSet segments) {
```
> TBH the method name also sounds too generic, since it only sends tombstone state.

`IracTombstoneManager#sendStateTo()` <- what kind of state have you thought about? 

yeah, I'll change the `origin` parameter's name.

danberindei replied:
Sorry, I wanted to copy your test in order to reply and I updated your comment instead :)

pruivo commented on core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java:
```diff
@@ -117,73 +151,104 @@ public int size() {
-         if (!info.isPrimary() || iracManager.running().containsKey(entry.getKey())) {
+         if (!info.isPrimary()) {
+            if (iracManager.running().containsKey(tombstone.getKey())) {
+               // key not sent yet
```
it won't send the update but it needs to keep track of it in case of primary owner crashes.

in this line, the node is looking for stale tombstones and, if the key exists in `IracManager`, it means the primary hasn't sent the update to the backup owners.

danberindei replied:
Got it, could you change the comment to something like "primary hasn't sent the update to the backup owners yet" for the benefit of my future self?

pruivo commented:
updated! added backpressure.

pruivo commented:
I don't think so.

pruivo commented:
rebased and updated:
* changed `org.infinispan.container.versioning.irac.DefaultIracTombstoneManager#sendStateTo()` to use `Flowable` instead of chaining the `CompletableFuture`

karesti reviewed (approved):
LVGTM
